File: sdks/typescript/src/resources/domains.ts
// Domains Resource for OpenAgentMail

import type { HttpClient } from '../utils/http.js';
import { paginate, PageIterator } from '../utils/pagination.js';
import type {
  Domain,
  AddDomainParams,
  ListDomainsParams,
  PaginatedResponse,
} from '../types.js';

/**
 * Client for managing domains
 */
export class DomainsClient {
  constructor(private readonly http: HttpClient) {}

  /**
   * Add a new domain
   */
  async add(params: AddDomainParams): Promise<Domain> {
    return this.http.post<Domain>('/domains', params);
  }

  /**
   * Get a domain by ID
   */
  async get(domainId: string): Promise<Domain> {
    return this.http.get<Domain>(`/domains/${domainId}`);
  }

  /**
   * Delete a domain
   */
  async delete(domainId: string): Promise<void> {
    return this.http.delete(`/domains/${domainId}`);
  }

  /**
   * Verify a domain
   */
  async verify(domainId: string): Promise<Domain> {
    return this.http.post<Domain>(`/domains/${domainId}/verify`);
  }

  /**
   * List domains with auto-pagination
   */
  list(params: ListDomainsParams = {}): PageIterator<Domain> {
    return paginate(
      (paginationParams) => this.listPage({ ...params, ...paginationParams }),
      params
    );
  }

  /**
   * List a single page of domains
   */
  async listPage(params: ListDomainsParams = {}): Promise<PaginatedResponse<Domain>> {
    const { podId, limit, pageToken } = params;
    return this.http.get<PaginatedResponse<Domain>>('/domains', {
      podId,
      limit,
      pageToken,
    });
  }
}

File: sdks/rust/src/resources/domains.rs
use std::sync::Arc;
use crate::client::ClientInner;
use crate::error::Result;
use crate::types::{AddDomainRequest, Domain, PaginatedResponse, PaginationParams};

#[derive(Debug, Clone)]
pub struct Domains { inner: Arc<ClientInner> }

impl Domains {
    pub(crate) fn new(inner: Arc<ClientInner>) -> Self { Self { inner } }

    pub async fn add(&self, request: AddDomainRequest) -> Result<Domain> {
        let response = self.inner.post("/domains").json(&request).send().await?;
        self.inner.handle_response(response).await
    }

    pub async fn list(&self, params: PaginationParams) -> Result<PaginatedResponse<Domain>> {
        let mut request = self.inner.get("/domains");
        for (key, value) in params.to_query_params() { request = request.query(&[(key, value)]); }
        let response = request.send().await?;
        self.inner.handle_response(response).await
    }

    pub async fn get(&self, domain_id: &str) -> Result<Domain> {
        let response = self.inner.get(&format!("/domains/{}", domain_id)).send().await?;
        self.inner.handle_response(response).await
    }

    pub async fn verify(&self, domain_id: &str) -> Result<Domain> {
        let response = self.inner.post(&format!("/domains/{}/verify", domain_id)).send().await?;
        self.inner.handle_response(response).await
    }

    pub async fn delete(&self, domain_id: &str) -> Result<()> {
        let response = self.inner.delete(&format!("/domains/{}", domain_id)).send().await?;
        self.inner.handle_empty_response(response).await
    }
}
